Add homekit pm type sensor

lesterlo opened the pull request:
Add pm2.5 and pm10 sensor type to the AirQualitySensor


## Breaking change
No Breaking Change


## Proposed change
The AirQualitySensor will now display PM2.5 Density and PM10 Density on the Homekit rather than Air Paticulate Density.


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:


```yaml
# Example configuration.yaml
sensor:
        - platform: template
          sensors:
                         pm25_sensor:
                                friendly_name: "PM 2.5 Sensor"
                                unique_id: "pm25_sensor"
                                value_template: "{{ state_attr('air_quality.pm_sensor', 'particulate_matter_2_5') }}"
                                unit_of_measurement: "μg/m³"
                         pm10_sensor:
                                friendly_name: "PM 10 Sensor"
                                unique_id: "pm10_sensor"
                                value_template: "{{ state_attr('air_quality.air_quality.pm_sensor', 'particulate_matter_10') }}"
                                unit_of_measurement: "μg/m³"
```

- This PR fixes or closes issue: fixes #
- This PR is related to issue:  https://community.home-assistant.io/t/air-quality-sensor-in-homekit-bridge/277060
- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum
  

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

lesterlo committed:
Add pm type sensor

Add pm2.5 and pm10 sensor type to the AirQualitySensor

homeassistant commented:
Hi @lesterlo, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#46060).

Once you do that we will be able to review and accept this pull request.

Thanks!

lesterlo committed:
Merge pull request #1 from home-assistant/dev

Update forked repo

lesterlo committed:
Merge branch 'dev' into add_pm_sensor_type

lesterlo reviewed (commented):
Update the code from home-assistant/core:dev HEAD in ce159d7db33cc494bc3038f71e340acea1922ab6

lesterlo committed:
Change variable name and fix code check problem

lesterlo committed:
change variable name and remove self

bdraco reviewed (changes requested):
Please add test coverage for the new accessory type

lesterlo commented:
> Please add test coverage for the new accessory type

Should I create a test in this file?
https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py

bdraco commented:
> > Please add test coverage for the new accessory type
> 
> Should I create a test in this file?
> [https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py](https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py?rgh-link-date=2021-02-07T08%3A27%3A34Z)

Yes please, and one to verify we properly create accessories with the device class in https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_get_accessories.py

lesterlo commented:
> > > Please add test coverage for the new accessory type
> > 
> > 
> > Should I create a test in this file?
> > [https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py](https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py?rgh-link-date=2021-02-07T08%3A27%3A34Z)
> 
> Yes please, and one to verify we properly create accessories with the device class in https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_get_accessories.py

Did I need to modify both file you mentioned to emulate a pm sensor for testing?
I am a newbie in home assistant development. Sorry for any inconvenience caused

bdraco commented:
https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py
Verify the sensor works the same way the PM sensor does.

https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_get_accessories.py
Verify the accessory can be created with the new device class.

lesterlo committed:
fix isort

lesterlo committed:
Add  pm10 & pm25 testing entity

lesterlo committed:
Fix the test error, readd the pm10 and pm25 sensor test

lesterlo commented:
Dear @bdraco,

Hi, I added a test coverage which are pm10 & pm25 sensor for the modification of homekit_controller/AirQualitySensor on [homekit/test_get_accessories.py](https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_get_accessories.py).

Since I modifiy some code inside the  homekit_controller/AirQualitySensor class, not creating a new accessories or sensor type. Would I need to modify [homekit/test_type_sensors.py](https://github.com/home-assistant/core/blob/dev/tests/components/homekit/test_type_sensors.py)? If yes, how to modify the `test_type_sensors.py` file to fulfill the test coverage?

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -164,6 +165,8 @@ def get_accessory(hass, driver, state, aid, config):
             a_type = "HumiditySensor"
         elif device_class == DEVICE_CLASS_PM25 or DEVICE_CLASS_PM25 in state.entity_id:
             a_type = "AirQualitySensor"
+        elif device_class == DEVICE_CLASS_PM10 or DEVICE_CLASS_PM10 in state.entity_id:
+            a_type = "AirQualitySensor"
```
```suggestion
        elif device_class in (DEVICE_CLASS_PM10, DEVICE_CLASS_PM25) or DEVICE_CLASS_PM25 in state.entity_id or DEVICE_CLASS_PM10 in state.entity_id:
            a_type = "AirQualitySensor"
```

bdraco replied:
Actually it would be better to create a new type for each class

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -139,13 +143,19 @@ def __init__(self, *args):
         """Initialize a AirQualitySensor accessory object."""
         super().__init__(*args, category=CATEGORY_SENSOR)
         state = self.hass.states.get(self.entity_id)
+        device_class = state.attributes.get(ATTR_DEVICE_CLASS)
```
We need a separate class for each type as `density_to_air_quality` will only map PM25
```
def density_to_air_quality(density):
    """Map PM2.5 density to HomeKit AirQuality level."""
    if density <= 35:
        return 1
    if density <= 75:
        return 2
    if density <= 115:
        return 3
    if density <= 150:
        return 4
    return 5

```

Make a base class of `AirQualitySensor` and have the new types inhert it

lesterlo replied:
So we need to modify the homekit component together with homekit-controller component?

After your review, I think we need to do the following modification

1. create a pm10 and pm25 class that inhert the AirQualitySensor class on both ` homekit` and ` homekit-controller` component.
2. create a new density_to_air_quality mapping for pm10.


Am I right?

bdraco reviewed (changes requested):
Please see comment about `density_to_air_quality` only supporting PM2.5

lesterlo committed:
Accept the reviewed code

Update homeassistant/components/homekit/accessories.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo commented:
@bdraco 

After your review, I think we need to do the following modification

1. create a pm10 and pm25 class that inhert the AirQualitySensor class on homekit component.
2. create a new density_to_air_quality mapping for pm10.


Am I right?

bdraco commented:
> @bdraco
> 
> After your review, I think we need to do the following modification
> 
> 1. create a pm10 and pm25 class that inhert the AirQualitySensor class on homekit component.
> 2. create a new density_to_air_quality mapping for pm10.
> 
> Am I right?

Sounds right to me.

bdraco reviewed (changes requested):
Please see comments above

lesterlo reviewed (commented):
I will find time to complete this taks.

lesterlo committed:
Merge branch 'dev' into add_pm_sensor_type

lesterlo committed:
Add a PM10 & PM25 class

Added PM10 and PM25 class that inherit the AirQualitySensor class.

Changed the density_to_air_quality mapping for pm10 and pm25

lesterlo commented:
Hello,


I created 2 class, pm10 & pm25 that inherit the base AirQualitySensor class.


I modify the density_to_air_quality mapping for pm 2.5 and pm 10

Pm2.5:
https://www.epa.vic.gov.au/for-community/environmental-information/air-quality/pm25-particles-in-the-air

PM10: 
https://www.epa.vic.gov.au/for-community/environmental-information/air-quality/pm10-particles-in-the-air

lesterlo committed:
Create an init function of the AirQualitySensor class

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -141,16 +149,22 @@ def __init__(self, *args):
         super().__init__(*args, category=CATEGORY_SENSOR)
         state = self.hass.states.get(self.entity_id)
+
+        self.init()
```
It looks like `AirQualitySensor` is no longer used?

lesterlo replied:
I think the AirQualitySensor can be treated as a generic class for non PM sensor. User can add any non PM 2.5 and PM10 air sensor to the homekit by using this class.

bdraco commented on homeassistant/components/homekit/util.py:
```diff
@@ -385,6 +385,32 @@ def density_to_air_quality(density):
+    return 5
+
+
+def density_to_air_quality_pm25(density):
```
There is a `density_to_air_quality_pm25` above that has the docstring `"""Map PM2.5 density to HomeKit AirQuality level."""` above but provides a different mapping.

lesterlo replied:
The `density_to_air_quality` is the original mapping function for the original`AirQualitySensor` class. The function should be mapped with a PM 2.5 value. I found a new mapping scale for [PM2.5](https://www.epa.vic.gov.au/for-community/environmental-information/air-quality/pm25-particles-in-the-air)and [PM10](https://www.epa.vic.gov.au/for-community/environmental-information/air-quality/pm10-particles-in-the-air). That's why I create a new mapping method for the PM2.5 and PM10.

If we remove the  `density_to_air_quality`, we could not retain the `AirQualitySensor` class for a generic air sensor. According to the Apple homekit menu, a homekit AirQualitySensor accessory class must provides a AIR_QUALITY value.

bdraco commented:
https://www.airnow.gov/sites/default/files/custom-js/conc-aqi.js

https://docs.airnowapi.org/aq101

https://www.airnow.gov/aqi/aqi-calculator-concentration/

bdraco commented:
I guess the next question is if the incoming data is averaged over 1 hour or 24 hours

lesterlo commented:
> I guess the next question is if the incoming data is averaged over 1 hour or 24 hours

I think it depends on you sensor design. Most of the sensors provide the current measurement data, not a averaged value. The mapping function is depend on the incoming data.

lesterlo committed:
Updated the test case for PM10 & PM2.5 Sensor

lesterlo committed:
Change the accessory name

lesterlo committed:
Merge branch 'home-assistant:dev' into add_pm_sensor_type

lesterlo committed:
Remove useless __init__ constructor

lesterlo commented:
@bdraco Any progress about this PR?

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -141,16 +149,22 @@ def __init__(self, *args):
         super().__init__(*args, category=CATEGORY_SENSOR)
         state = self.hass.states.get(self.entity_id)
+
+        self.init()
```
```suggestion
        self.create_services()
```

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -141,16 +149,22 @@ def __init__(self, *args):
+        # GET to avoid an event storm after homekit startup
+        self.async_update_state(state)
+
+    def init(self):
```
```suggestion
    def create_services(self):
```

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -166,6 +180,58 @@ def async_update_state(self, new_state):
+class PM10Sensor(AirQualitySensor):
+    """Generate a PM10Sensor accessory as PM 10 sensor."""
+
+    def init(self):
```
```suggestion
    def create_services(self):
```

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -166,6 +180,58 @@ def async_update_state(self, new_state):
+class PM25Sensor(AirQualitySensor):
+    """Generate a PM25Sensor accessory as PM 2.5 sensor."""
+
+    def init(self):
```
```suggestion
    def create_services(self):
```

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -166,6 +180,58 @@ def async_update_state(self, new_state):
+    def async_update_state(self, new_state):
+        """Update accessory after state change."""
+        density = convert_to_float(new_state.state)
+        if density:
```
```suggestion
        if not density:
            return
```

Then outdent the block below

lesterlo replied:
Why do we need to return for "if not density:" here?

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -166,6 +180,58 @@ def async_update_state(self, new_state):
+    def async_update_state(self, new_state):
+        """Update accessory after state change."""
+        density = convert_to_float(new_state.state)
+        if density:
```
```suggestion
        if not density:
            return
```

Then outdent the block below

bdraco reviewed (changes requested):
Looks good now. 

We just need tests for PM10 and PM25

Likely duplicate this test of tests and write them for pm10 and pm25
```
async def test_air_quality(hass, hk_driver):
```

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

frenck commented:
@lesterlo If you need help with the test, please shout out! Maybe we can help in giving directions to move this PR forward 👍

lesterlo commented:
> @lesterlo If you need help with the test, please shout out! Maybe we can help in giving directions to move this PR forward 👍

Dear @frenck, 

Where should I create a new test for pm10 and pm25?

```async def test_air_quality(hass, hk_driver):```

lesterlo commented:
Any thing I need to do before merging?

bdraco commented:
I think we just need the test https://github.com/home-assistant/core/pull/46060#pullrequestreview-707877717

lesterlo commented:
@bdraco 

How to test? Do we need to write some code for testing?

matthewhadley commented:
@lesterlo I believe this:

> Likely duplicate this test of tests and write them for pm10 and pm25
> 
> ```
> async def test_air_quality(hass, hk_driver):
> ```

Is referring to this:

https://github.com/home-assistant/core/blob/332a571bb480ccfa5dd42068346986099c0b1c5d/tests/components/homekit/test_type_sensors.py#L94

lesterlo commented:
Dear @matthewhadley ,
OK, let me take a look at this. Thanks.

lesterlo commented:
Let me try to finish it or make some progress on this weekend.

lesterlo committed:
Add the pm10 and pm25 sensor tests

lesterlo committed:
Merge branch 'home-assistant:dev' into add_pm_sensor_type

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -178,6 +192,60 @@ def async_update_state(self, new_state):
+        density = convert_to_float(new_state.state)
+        if not density:
+            return
+            if self.char_density.value != density:
+                self.char_density.set_value(density)
+                _LOGGER.debug("%s: Set density to %d", self.entity_id, density)
+            air_quality = density_to_air_quality_pm10(density)
+            if self.char_quality.value != air_quality:
+                self.char_quality.set_value(air_quality)
+                _LOGGER.debug("%s: Set air_quality to %d", self.entity_id, air_quality)
```
It looks like these lines need to be outdented

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -178,6 +192,60 @@ def async_update_state(self, new_state):
+        density = convert_to_float(new_state.state)
+        if not density:
+            return
+            if self.char_density.value != density:
+                self.char_density.set_value(density)
+                _LOGGER.debug("%s: Set density to %d", self.entity_id, density)
+            air_quality = density_to_air_quality_pm25(density)
+            if self.char_quality.value != air_quality:
+                self.char_quality.set_value(air_quality)
+                _LOGGER.debug("%s: Set air_quality to %d", self.entity_id, air_quality)
```
```suggestion
        if self.char_density.value != density:
            self.char_density.set_value(density)
            _LOGGER.debug("%s: Set density to %d", self.entity_id, density)
        air_quality = density_to_air_quality_pm25(density)
        if self.char_quality.value != air_quality:
            self.char_quality.set_value(air_quality)
            _LOGGER.debug("%s: Set air_quality to %d", self.entity_id, air_quality)
```

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lesterlo committed:
Update homeassistant/components/homekit/type_sensors.py

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco reviewed (changes requested):
Please resolve conflicts.  Thanks 👍

lesterlo commented:
> Please resolve conflicts. Thanks 👍

ok, Let me fix it on this week

lesterlo committed:
Merge branch 'dev' into add_pm_sensor_type

lesterlo commented:
Hello, I am just fixed the conflicts

bdraco commented:
Started the CI. Will come back and check the results in a bit

lesterlo committed:
Fix the tests expected result

lesterlo commented:
Hello, I have fixed the test case expected result of PM10 and PM25 sensor. Please re-run the CI.

bdraco commented:
CI started

lesterlo committed:
Complete the testcase for pm10 and pm25 sensor
